src/MergeContext.js

```diff
@@ -454,6 +454,202 @@ class BranchPosition
+        assert(this._author !== null && this._body !== null && this._trailer !== null);
+        let message = this._title;
+        if (this._body)
+            message += '\n\n' + this._body;
+        if (this._trailer)
+            message += '\n\n' + this._trailer;
+        return message;
+    }
+
+    // returns author object in the {name, email, date} format
+    createAuthor(mergeCommitAuthor)
+    {
+        if (!this._author)
+            return mergeCommitAuthor;
+        return {name: this._author.name, email: this._author.email, date: mergeCommitAuthor.date};
+    }
+
+    author() { return this._author; }
+
+    // returns the position of the first non-ASCII_printable character (or -1)
+    _invalidCharacterPosition(str) {
+        const prohibitedCharacters = /[^\u{20}-\u{7e}]/u;
+        const match = prohibitedCharacters.exec(str);
+        return match ? match.index : -1;
+    }
+
+    // basic checks for the entire raw message
+    _checkMessage() {
+        // constructor removed CRs in CRLF sequences
+        // other CRs are not treated specially (and are banned)
+        const lines = this.whole().split('\n');
+        for (let i = 0; i < lines.length; ++i) {
+            const line = lines[i];
+            const invalidPosition = this._invalidCharacterPosition(line);
+
+            if (invalidPosition !== -1)
+                throw new Error(`bad character at line ${i}, offset ${invalidPosition}`);
+
+            if (line.length > 72)
+                throw new Error(`too long line '${line}'`);
+        }
+    }
+
+    // checks whether the message has a misused attribute
+    _checkAttributes() {
+        // A line matched by this expression is considered valid if it
+        // starts with one of the allowed attributes and invalid otherwise.
+        // A line that is not matched by this expression and containing one
+        // of the allowed attributes is ignored.
+        const authorExp = /^\s*\S*[aA]uthored[-_]?[bB]y/;
+        let lines = this._body.split('\n');
+        while (lines.length > 0) {
+            const line = lines.shift();
+            if (line.match(authorExp)) {
+                if (this._attributes.some(attr => line.startsWith(attr + ':')))
+                    continue;
+                throw new Error(`a misused '*Authored-by' attribute in the message`);
+            }
+        }
+    }
+
+    // parses attributes of the raw PR message
+    _parse() {
+        try
+        {
+            this._checkMessage();
+
+            this._checkAttributes();
+
+            // index of the last occurrence of '\n\n'
+            const trailerIndex = this._body.search(/\n\n(?!.+\n\n)/s);
+            if (trailerIndex >= 0) {
+                this._trailer = this._body.substring(trailerIndex).trim();
+                this._body = this._body.substring(0, trailerIndex); // trim in _parseBody()
+                this._parseTrailer();
+            }
+
+            this._parseBody();
+
```

Comment:
The proposed parseTrailer(), parseBody/Header() order does not feel natural. If possible, please refactor using this high-level sketch:

```js
        this.extractHeader();

        try {
           this.extractTrailer();
        }
        catch (e) {
             log(... assuming no trailer: e ...);
         }

        this._checkForTypos();
 ```

where extractHeader() is something along these lines:

```js
        if (this._body.startsWith('Authored-by:')) {
            const tokenizer = new FieldsTokenizer(this._body);
            const authorFieldValue = tokenizer.nextField('Authored-by');
            this._author = this._parseAuthor(authorFieldValue);
            if (tokenizer.nextOptionalField())
                throw ... headers can only have one (Authored-by) field ...
            this._body = tokenizer.remaining();
        }
```